pkg/zkproofs/ciphertext_validity.go

```diff
@@ -0,0 +1,163 @@
+)
+
+// CiphertextValidityProof represents a zero-knowledge proof that a ciphertext is valid.
+type CiphertextValidityProof struct {
+	Commitment1 curves.Point
+	Commitment2 curves.Point
+	Challenge   curves.Scalar
+	Response1   curves.Scalar
+	Response2   curves.Scalar
+}
+
+// NewCiphertextValidityProof generates a zero-knowledge proof that a ciphertext is properly encrypted.
+func NewCiphertextValidityProof(r *curves.Scalar, pubKey curves.Point, ct *elgamal.Ciphertext, message uint64) (*CiphertextValidityProof, error) {
+	// Validate input
+	if r == nil {
+		return nil, errors.New("invalid randomness factor")
+	}
+
+	if pubKey == nil {
+		return nil, errors.New("invalid public key")
+	}
+
+	if ct == nil || ct.C == nil || ct.D == nil {
+		return nil, errors.New("invalid ciphertext")
+	}
+
+	eg := elgamal.NewTwistedElgamal()
+
+	H := eg.GetH()
+	G := eg.GetG()
+
+	ed25519 := curves.ED25519()
+	// Convert message to a scalar
+	messageValue := new(big.Int).SetUint64(message)
+	x, _ := ed25519.Scalar.SetBigInt(messageValue)
+
+	// Step 1: Generate random blinding factors for the proof
+	rBlind := ed25519.Scalar.Random(crand.Reader) // Blinding factor for random value r
+	xBlind := ed25519.Scalar.Random(crand.Reader) // Blinding factor for random value x
+
+	// Step 2: Create commitments
+	rBlindH := H.Mul(rBlind)            // rBlind * H
+	xBlindG := G.Mul(xBlind)            // xBlind * G
+	Commitment1 := rBlindH.Add(xBlindG) // Commitment1 = rBlind * H + xBlind * G
+
+	Commitment2 := pubKey.Mul(rBlind) // Commitment2 = rBlind * P
+
+	// Step 3: Generate a challenge using the Fiat-Shamir heuristic.
+	// The challenge is basically just a hash of all the provided values. This locks in the values and makes sure that
+	// the proof cannot be for some other set of values.
+	hashData := append(Commitment1.ToAffineCompressed(), Commitment2.ToAffineCompressed()...)
+	hashData = append(hashData, ct.C.ToAffineCompressed()...)
+	hashData = append(hashData, ct.D.ToAffineCompressed()...)
+	challenge := ed25519.Scalar.Hash(hashData)
+
+	// Step 4: Generate responses
+	Response1 := challenge.MulAdd(*r, rBlind) // Response1 = rBlind + challenge * r
+	Response2 := challenge.MulAdd(x, xBlind)  // Response2 = xBlind + challenge * x
+
+	return &CiphertextValidityProof{
+		Commitment1: Commitment1,
+		Commitment2: Commitment2,
+		Challenge:   challenge,
+		Response1:   Response1,
+		Response2:   Response2,
+	}, nil
+}
+
+// VerifyCiphertextValidityProof verifies the zero-knowledge proof that a ciphertext is valid.
+func VerifyCiphertextValidityProof(proof *CiphertextValidityProof, pubKey curves.Point, ct *elgamal.Ciphertext) bool {
+	// Validate input
+	if proof == nil || proof.Commitment1 == nil || proof.Commitment2 == nil || proof.Challenge == nil ||
+		proof.Response1 == nil || proof.Response2 == nil || pubKey == nil || ct == nil || ct.C == nil || ct.D == nil {
+		return false
+	}
+
+	eg := elgamal.NewTwistedElgamal()
+	H := eg.GetH()
+	G := eg.GetG()
+
```

Comment:
We should add a step here to recompute the Challenge scalar as well:
This should be done in the same way as the proof generation so we could create a helper function for it:

hashData := append(proof.Commitment1.ToAffineCompressed(), proof.Commitment2.ToAffineCompressed()...)
hashData = append(hashData, ct.C.ToAffineCompressed()...)
hashData = append(hashData, ct.D.ToAffineCompressed()...)
challenge := ed25519.Scalar.Hash(hashData)